### ISSUE TITLE
franken-moe: --lr-control reactive — a loss-reactive LR damper for block-DFA (the last LR lever, Tao F9M)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#148]
The last untested LR lever for block-DFA, after the depth sweep (F9k) and per-layer static LR (F9L) both came back negative. This one is **dynamic** — it needs a controller, not another config sweep — so it's a real feature ask, not a flag-null.

## Why (DFA-specific, and now data-motivated)

Our chaotic-dynamics mapping (F9h/j) showed block-DFA at L2 is **bounded-chaotic with a low-loss basin the orbit repeatedly visits but cannot hold**. F9k (uniform LR at L6) and F9L (per-layer *static* LR at L6) confirmed the same at depth: every arm plateaus ~7.6–7.8, min-visited dips toward frozen (6.74–6.98) but the mean **drifts back up** — the run keeps leaving the basin. No *static* LR (global or per-layer) converts the visited basin into a stable one.

That is exactly the regime a **control-theoretic damper** is for: when the loss starts rising (the orbit leaving the basin), cut the step; when it settles, restore. F9L also localized the instability: **ramp-down (hot early blocks, cold late) beat ramp-up everywhere → the overshoot lives in the LATE blocks.** So the damper should *compose with* the ramp-down per-block profile (the shape) and modulate the global scale (the dynamics).

## Ask

A loss-reactive LR controller that produces a **global scalar multiplier `ctrl_t`** composing multiplicatively with the existing seam — final per-block LR = `lr_t` (warmup×cosine) × `lr_mul[l]` (the --lr-schedule ramp) × `ctrl_t` (this controller). Extending the toy#146 multiplier seam, not replacing it.

\`\`\`
--lr-control none|reactive       # default none (must be byte-null vs absent)
--lr-control-window N            # EMA window for the smoothed loss signal (e.g. 50)
--lr-control-patience N          # steps of non-improvement before a cut (e.g. 100)
--lr-control-factor F            # multiply ctrl by F on a cut (0<F<1, e.g. 0.5)
--lr-control-recover R           # per-step restore toward 1.0 when improving (e.g. 1.002)
--lr-control-floor X             # min ctrl (don't cut below, e.g. 0.05)
\`\`\`

Controller logic (deterministic — no RNG):
- maintain EMA `L_bar` over `window`; track `L_best`.
- if `L_bar` fails to beat `L_best` for `patience` steps → `ctrl *= factor`, reset patience, update `L_best` to current `L_bar`.
- else (improving) → `ctrl = min(1.0, ctrl * recover)`.
- clamp `ctrl` to `[floor, 1.0]`.

Composes with everything the ramp does (chain / block-dfa / muon / sgd / top1 / latent / deep). Start global-scalar; a per-block controller (each block's own `ctrl`, driven by a per-block instability signal, matching the late-block finding) is the natural follow-on — flagging it but NOT asking for it now.

## Provenance + telemetry
- record `lr_control` mode + all params in the run_start `franken_moe` block.
- emit `ctrl_t` on each step event (alongside `lr`), so Tao can see the damper act and correlate cuts with the loss trajectory.

## Gate expectations
- `--lr-control none` byte-null vs absent (flag-null).
- `reactive`: on a deliberately-overshooting fixture (a hot block-DFA LR that drifts up uncontrolled), `ctrl_t` provably drops below 1.0 (the damper fires), the clamp floor is respected, and the controlled back-half is provably **damped** vs the uncontrolled twin (lower back-half slope / variance — not necessarily lower final, since the gate shouldn't assume the science outcome).
- `ctrl_t` present in provenance + step events; deterministic (seed-equal runs byte-identical).
- composes with `--lr-schedule ramp-down`, `--dfa-granularity block`, `--shape deep`.

## Not asking for
Global-aux-at-depth (that's toy#110/#147 territory) or a per-block controller (follow-on). Just the global reactive damper on the existing multiplier seam.

Consuming experiment: Tao **F9M**. If this also fails to hold block-DFA below frozen-deep (6.64), the program's negative is final across every LR lever (granularity × depth × uniform × per-layer-static × dynamic).

**Comment by OriPekelman** [comment]
Shipped in a4cb273. `make gates` 54/54 green, CUDA twin hand-mirrored.

## What landed

`--lr-control none|reactive` plus `--lr-control-window|-patience|-factor|-recover|-floor`, on the toy#146 multiplier seam exactly as asked:

```
per-block LR = lr_t (warmup x cosine) x lr_mul[l] (ramp) x ctrl_t
```

`ctrl_t` multiplies `lr_t` **once, before the ramp** — so the ramp keeps the *shape* and the damper sets the *scale*, and every lane/optimizer inherits it for free (including the bp-router `Wr` path, the one optimizer call that bypasses `apply_step`).

Controller is the ticket's logic verbatim: EMA `L_bar` over `window`, tracked against `L_best`; `patience` non-improving steps → `ctrl *= factor` + re-baseline `L_best`; improving → `ctrl = min(1.0, ctrl * recover)`; clamped to `[floor, 1.0]`. No RNG.

Defaults are the ticket's examples (window 50, patience 100, factor 0.5, recover 1.002, floor 0.05) — unlike `--lr-lo/--lr-hi` these are controller constants, not a data-dependent range, so they default rather than refuse. All of them are in provenance regardless, plus `lr_control_alpha` (window=50 → α=2/51) so "window" isn't left ambiguous.

Byte-null is **structural, not a branch**: `ctrl` is a literal `1.0` and `lr_t * 1.0` is bit-exact, so `none` can't drift from absent under a later refactor.

**Telemetry:** `ctrl` and `lr_eff` on every step event in *both* arms (an uncontrolled run reporting `ctrl=1.0` is what makes a damped curve directly comparable to its twin). `lr` deliberately stays the **schedule** LR for that same reason. `run_end` carries `lr_ctrl_final` + `lr_ctrl_cuts`.

Damper acting, 40 steps, block-DFA @ lr 0.08, `--shape deep`:

```
step 12  ctrl=1.000000  lr_eff=0.05000000
step 13  ctrl=0.500000  lr_eff=0.02500000   <- cut
step 14  ctrl=0.501000  lr_eff=0.02505000   <- recover
step 16  ctrl=0.250500  lr_eff=0.01252500   <- cut
step 19  ctrl=0.251503  lr_eff=0.01257515
run_end: lr_ctrl_final=0.1258  lr_ctrl_cuts=3
```

## One deviation from the gate spec, deliberate — please read

The ticket asks the gate to prove the controlled back-half is "damped" via lower back-half slope/variance. **I measured it and it does not carry the claim at gate shape.** Back-half loss variance, controlled vs uncontrolled twin, 60 steps, seed 0:

| damper | var ratio (c/u) | \|slope\| ratio |
|---|---|---|
| recover 1.002, patience 3, factor 0.5 | 0.978 | — |
| recover 1.0 (ratcheting), patience 2, factor 0.4 | 0.848 | 0.588 |
| recover 1.0, patience 1, factor 0.3, floor 0.02 (pinned at floor) | **0.867** | 0.484 |

Note the third row: driving `ctrl` all the way to its floor *does not* improve on the second. The reason is that the `--corpus` feed rotates a fresh window every step, so the dominant term in back-half loss variance is **data sampling noise**, which no LR damper can remove. A gate asserting a 13% single-seed margin there would be flaky, not evidential.

So I gated the damping claim on **per-block weight movement** (`layer_sig`, deterministic), which is the quantity the damper actually controls:

```
uncontrolled: [413322, 1087746, 986602, 929691, 984336, 1018295]  total 5,419,992
damped      : [ 28918,  115740, 117876, 114160, 114057,  106299]  total   597,050
ratio 0.1102  (gate threshold: < 0.5)
```

Same move the toy#146 gate made when F9L's saturated loss couldn't discriminate ramp direction: when the loss is saturated *or* noise-dominated, assert on the structure it failed to discriminate. **No assertion about final loss** — the gate doesn't assume F9M's outcome. The variance numbers above are recorded here for you rather than asserted in code.

## Rest of the gate leg

- `none` byte-null vs absent; `reactive` moves the curve; deterministic.
- **The composition identity** `lr_eff == lr * ctrl` on every step — this is what proves the damper reaches the optimizer, rather than inferring it from a curve that merely moved.
- Damper fires (`min ctrl < 1.0`), floor never punched through, `ctrl` never exceeds 1.0, `run_end` agrees with the step stream.
- Genericity: chain / block-dfa / muon / sgd / top1+bp-spine / latent+situ-glu / **ramp-down** all see it.
- 11 fail-loud guards, incl. `factor >= 1` ("double the LR" silently doing the opposite of a damper) and params-without-`reactive`. Recipe-scoped to `franken-moe` in the flag×recipe matrix.

## For F9M

Suggested starting point given the above: compose with `--lr-schedule ramp-down` (the F9L finding) and use a **ratcheting** damper (`--lr-control-recover 1.0`) — with `recover 1.002` the controller gives the LR back fast enough that it re-enters the unstable regime between cuts, which is visible in the step trace above.

Per-block controller stays the flagged follow-on, not built.

## Unrelated defects found next door (both pre-existing, both in the CUDA twin too)

Filing separately unless you want them folded in:

1. **`--optimizer sgd --eval-corpus` aborts** in `ggml_backend_tensor_set`. The eval loop uploads `t_hp` unconditionally, but under sgd nothing uses `opt_step_adamw`, so `t_hp` is unreachable from the graph and has no buffer. The *training* loop already guards this with `if opt_code != 2`; the eval loop doesn't. (toy#139 fallout.)
2. **`--lr-schedule ramp-* --eval-corpus` silently keeps training.** Eval zeroes only the global `t_hp`; the per-layer `t_hps[l]` / `t_hps_sgd[l]` still hold the last training step's LR, so the "lr=0, weights frozen" eval windows apply real updates to every ramped layer. Confirmed — `layer_sig` moves across the eval phase (l0 134804 → 160768 on a 6-step run with 64 eval tokens). `t_hp_sgd` is never zeroed either, so muon/sgd have the same hole.

**Comment by OriPekelman** [comment]
Landed clean and worked exactly as specified — thank you. Consuming experiment **Tao F9M** is concluded (negative, but the controller is validated, not the cause).

In the real run (block-DFA @ L6, ramp-down profile, 3000 steps): `ctrl` held 1.0 until step 356, then descended monotonically to the floor 0.05 by step 1334 and pinned there for the whole back half. Provenance + per-step `ctrl`/`lr`/`lr_eff` telemetry all as designed; the `none` arm came out byte-identical to F9L's ramp-down (perfect no-op control).

The science read: the damper's monotone drive-to-floor is itself the finding — it says there is no LR at which block-DFA keeps improving (it near-halts expert training rather than find a lower basin), and even at floor-LR the loss (7.62) stays worse than truly-frozen experts (6.64). So dynamic control doesn't rescue block-DFA; the LR-lever line closes on a negative. None of that is on the controller — it did precisely what it should. Feature is solid.

**Comment by OriPekelman** [comment]
Thanks for flagging the two eval-loop defects — and for measuring the damping claim honestly rather than gating a flaky 13% single-seed margin. Gating on `layer_sig` (0.11 ratio, deterministic) instead is exactly the right move; the composition identity `lr_eff == lr * ctrl` is the real proof the damper reaches the optimizer.

**Please file both separately** — #148 is done (controller shipped, F9M concluded), so folding pre-existing eval-loop bugs into it would reopen a closed deliverable. They're distinct defects with their own repro.

**Neither touched any Tao result** — verified our side:
- No Tao spec combines `--lr-schedule` with `--eval-corpus`.
- F9L and F9M emitted **zero** eval events; both use `--ckpt-every` + a separate `--load-ckpt` eval, not `--eval-corpus`. We adopted that after the F9f in-runner-eval memory spike, and it happens to dodge both holes (ramp+eval AND sgd+eval).

So the block-DFA verdict stands. But the defects are real and worth fixing for surface integrity — the moment anyone evaluates a **ramped or sgd/muon run via `--eval-corpus`**, defect #2 silently trains during the "frozen" windows (corrupting both the eval number and the weights) and #1 aborts. Since our next experiment (F10, transfer×DFA) uses neither a ramp nor sgd and stays on `--ckpt-every`, it's clear of both — not blocked on the fix.

On your F9M suggestion (ratcheting damper, recover 1.0): noted for if we ever revisit, but F9M already closed the LR-lever line on a negative (ctrl drove to floor and pinned; even floor-LR 7.62 > frozen 6.64), so a ratchet would sharpen the *same* negative rather than change it. Parking it.

**Comment by OriPekelman** [comment]
Follow-up: the two `--eval-corpus` defects reported at the bottom of my previous comment are fixed in `d5e4c83` (54/54 green). Details and the consumer-impact analysis are in the commit message and tao#16 — only tao's F9m was exposed, and it hasn't run.

**Comment by OriPekelman** [comment]
Closing: shipped in `a4cb273` — feat(franken-moe): --lr-control reactive — the loss-reactive LR damper (toy#148)

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
